aws-xray-recorder-sdk-sql/src/main/java/com/amazonaws/xray/sql/TracingStatement.java

```diff
@@ -0,0 +1,134 @@
+                //invoke the original method "wrapped" in a XRay Subsegment
+                //implementation note : can not use shorter AWSXRay.createSubsegment because we need to deal with InvocationTargetException
+                boolean openDummySegment = false;
+                if (! AWSXRay.getCurrentSegmentOptional().isPresent()) {
+                    //we are (probably) being executed outside AWS env (unit tests, background process, ...)
+                    // -> we do not want to crash or to log a SegmentNotFoundException, so we print to stdout
+                    // note that if you are in AWS context, stdout will be log to cloudwatch logs anyway
+                    openDummySegment = true;
+                    System.out.println(sanitizeSql(resolveExecutedSql(method.getName(), args)));
+                    AWSXRay.beginDummySegment();
+                }
+                Subsegment subsegment = AWSXRay.beginSubsegment("SQL");
+                subsegment.putAllSql(extractSqlParams(method, args));
+                subsegment.setNamespace(Namespace.REMOTE.toString());
+                try {
+                    return method.invoke(original, args); //execute the query
+                } catch (Throwable t) {
+                    subsegment.addException(t);
+                    if (t instanceof InvocationTargetException && t.getCause() != null) {
+                        throw t.getCause();
+                    } else {
+                        throw t;
+                    }
+                } finally {
+                    AWSXRay.endSubsegment();
+                    if (openDummySegment) {
+                        AWSXRay.endSegment();
+                    }
+                }
+
+            } else {
+                //else, simply delegates
+                return method.invoke(original, args);
+            }
+        }
+
+        private Map<String, Object> extractSqlParams(Method method, Object[] args) {
+            Map<String, Object> additionalParams = new HashMap<>();
+            try {
+                DatabaseMetaData metadata = original.getConnection().getMetaData();
+                additionalParams.put("url", metadata.getURL());
+                additionalParams.put("driver_version", metadata.getDriverVersion());
+                additionalParams.put("database_type", metadata.getDatabaseProductName());
+                additionalParams.put("database_version", metadata.getDatabaseProductVersion());
+                additionalParams.put("sanitized_query", sanitizeSql(resolveExecutedSql(method.getName(), args)));
+            } catch (SQLException e) {
+                //a SqlException at this stage is kind of strange ... just ignore
+                // (i.e the XRay trace will not contain the info but this is not very severe)
+                //at least, we provide a way to identify the problem in the XRay trace
+                additionalParams.put("url", "strange SQLException occurred ...");
```

Comment:
I think we should log it instead of putting it inside the segment.

Since this is sort of a catch_all for a SQL Exception, it would be more informative to be able to see the stack trace with the logger as well. 
`logger.warn("add informative message here", e)` to directly include the stack trace.

Reply:
For MySQL plugin we use
```
logger.warn("Unable to create statement proxy for tracing.", e);
```